File: ruby/robot-name/robot_name.rb
class Robot
  attr_reader :name

  @@taken_names = []

  def initialize
    @name = generate_random_name
  end

  def reset
    @@taken_names.delete(name)
    @name = generate_random_name
  end

  private

  VIABLE_LETTERS = ('A'..'Z').to_a

  def generate_random_name
    name = "#{two_random_letters}#{three_random_numbers}"
    if @@taken_names.include?(name)
      generate_random_name
    else
      @@taken_names << name
      name
    end
  end

  def two_random_letters
    "#{VIABLE_LETTERS.sample}#{VIABLE_LETTERS.sample}"
  end

  def three_random_numbers
    "#{rand(1..9)}#{rand(1..9)}#{rand(1..9)}"
  end
end

module BookKeeping
  VERSION = 2
end

File: javascript/robot-name/robot-name.js
const takenNames = {};

const viableLetters = [
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
  'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
  'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
];

const twoRandomLetters = () => {
  return (
    `${viableLetters[Math.floor(Math.random() * viableLetters.length)]}` +
    `${viableLetters[Math.floor(Math.random() * viableLetters.length)]}`
  );
};

const threeRandomNumbers = () => {
  return (
    `${Math.floor(Math.random() * 10)}` +
    `${Math.floor(Math.random() * 10)}` +
    `${Math.floor(Math.random() * 10)}`
  );
};

class Robot {
  constructor(){
    this.name = this.generateRandomName();
  }

  generateRandomName(){
    const name = `${twoRandomLetters()}${threeRandomNumbers()}`;
    if (takenNames[name]) {
      this.generateRandomName();
    } else {
      takenNames[name] = name;
      return name;
    }
  }

  reset(){
    this.name = this.generateRandomName();
  }

}

module.exports = Robot;
